miden/examples/nprime/nprime.masm

```diff
@@ -0,0 +1,95 @@
+# push the current prime+count tuple to the memory
+proc.push
+    # define the index of the prime to be added
+    dup.1
+
+    # push the current candidate to the primes list
+    mem_store
+end
+
+# push the next prime to the stack with the primes count
+proc.next
+    # initial state
+    # [last prime, primes count]
+
+    # push candidate & not found flag
+    # [not found, candidate, last prime, primes count]
+    dup
+    push.1
+
+    while.true
+        # increment the candidate
+        # [accumulator, candidate, last prime, primes count]
+        add.1
+        push.0
+
+        # push counters
+        # [i < n, i, n, accumulator, candidate, last prime, primes count]
+        dup.3
+        push.0
+        push.1
+
+        while.true
+            # fetch the candidate
+            # [candidate, i, n, accumulator, candidate, last prime, primes count]
+            dup.3
+
+            # fetch the current prime
+            # [prime, candidate, i, n, accumulator, candidate, last prime, primes count]
+            dup.1
+            add.1
+            mem_load
+
+            # calculate the mod prime
+            # this is a naive scan for the sake of simplicity. normally we should check up to sqrt of prime
+            # [remainder, i, n, accumulator, candidate, last prime, primes count]
+            u32checked_mod
+
+            # check if is divisor and update the accumulator
+            # [i, n, accumulator, candidate, last prime, primes count]
+            u32checked_eq.0
+            dup.3
+            add
+            swap.3
+            drop
+
+            # increment i up to n
+            add.1
+            dup
+            dup.2
+            u32checked_lt
+        end
+
+        # loop finished. if accumulator > 0, then the number has at least one divisor
+        # [accumulator, candidate, last prime, primes count]
+        drop
+        drop
+        u32checked_neq.0
+    end
+
+    # increment the prime counter
+    movup.2
+    add.1
+
+    # return the calculated prime
+    swap
+    exec.push
+end
+
+begin
```

Comment:
Would be great to add a comment here to describe what is the expected state of the stack at the start of program execution.

Also, and this is completely optional, would be great to be able to specify $n$ dynamically via stack inputs. This would require some modification to the while loop below.

Reply:
Perfect, onto that!

Reply:
I modified the example to not use a naive approach anymore.

The only sequential primes are 2 & 3. Provided we treat this case as exception, we can iterate with windows of `2`.

Also, we are now benefiting from the algebraic property of composite numbers that its smallest prime factor squared will be less than or equal to the composite number itself. This means we should iterate only to the square of smallest factor - otherwise, the number necessarily is a prime.

Reply:
I also added documentation for the main procedure, with an example of input/output

Reply:
We are now calculating the 1000th prime in 2 seconds:

```
⚡ cargo run --release --features executable --bin miden -- run -a examples/nprime/nprime.masm
    Finished release [optimized] target(s) in 0.04s
     Running `/home/vlopes/dev/polygon/miden/target/release/miden run -a examples/nprime/nprime.masm`
============================================================
Run program
============================================================
Reading program file `examples/nprime/nprime.masm`
Compiling program... done (1 ms)
Reading input file `examples/nprime/nprime.inputs`
Executing program with hash 8ad9c29d910e183a6590ae0a4f94f7cb2453369e8e556e304f
4e0d35b078f096... done (2065 ms)
Output: [7919, 7907, 7901, 7883, 7879, 7877, 7873, 7867, 7853, 7841, 7829, 7823, 7817, 7793, 7789, 7759]
```